modules/index-service/src/main/java/org/opencastproject/index/service/impl/util/PublicationEventUpdateHandler.java

```diff
@@ -0,0 +1,203 @@
+      // We don't want to handle anything but TakeSnapshot messages.
+      return;
+    }
+    AssetManagerItem.TakeSnapshot snapshotItem = (AssetManagerItem.TakeSnapshot) messageItem;
+    updatePublications(snapshotItem.getMediapackage());
+  }
+
+  private void updatePublications(MediaPackage mediaPackage) {
+    for (Publication publication : mediaPackage.getPublications().clone()) {
+      if (!configurationPublicationChannelIds.contains(publication.getChannel())) {
+        continue;
+      }
+      String channelId = publication.getChannel();
+      String publicationId = publication.getIdentifier();
+      URI publicationURI = publication.getURI();
+      MimeType publicationMimeType = publication.getMimeType();
+
+      if (channelId == null || publicationURI == null || publicationMimeType == null || publicationId == null) {
+        // No publication, do nothing
+        continue;
+      }
+
+      // Get metadata and ACLs
+      SimpleElementSelector elementSelector = new SimpleElementSelector();
+      for (String flavor : sourceFlavors) {
+        elementSelector.addFlavor(flavor);
+      }
+      Collection<MediaPackageElement> elements = elementSelector.select(mediaPackage, false);
+
+      Set<String> elementIds = new HashSet<>();
+      for (MediaPackageElement elem : elements) {
+        elementIds.add(elem.getIdentifier());
+      }
+
+      if (elementIds.size() < 1) {
+        // Nothing to republish, do nothing
+        continue;
+      }
+
+      // Remove publication from mediapackage
+      mediaPackage.remove(publication);
+
+      List<MediaPackageElement> downloadElements;
+      List<MediaPackageElement> streamingElements = new ArrayList<>();
+      try {
+        downloadElements = downloadDistributionService.distributeSync(channelId, mediaPackage, elementIds, false);
+        if (streamingDistributionService != null && streamingDistributionService.publishToStreaming()) {
+          streamingElements = streamingDistributionService.distributeSync(channelId, mediaPackage, elementIds);
+        }
+      } catch (DistributionException e) {
+        throw new RuntimeException(e);
+      }
+      logger.debug("Distribute of mediapackage {} to channel {} completed", mediaPackage, channelId);
+
+      // Re-add publication
+      Publication newPublication = PublicationImpl.publication(publicationId, channelId, publicationURI, publicationMimeType);
+
+      // Add published elements
+      for (MediaPackageElement element : downloadElements) {
+        element.setIdentifier(null);
+        PublicationImpl.addElementToPublication(newPublication, element);
+      }
+      for (MediaPackageElement element : streamingElements) {
+        element.setIdentifier(null);
+        PublicationImpl.addElementToPublication(newPublication, element);
+      }
+
+      mediaPackage.add(newPublication);
+    }
+
+    // Also search service
+    try {
+      searchService.addSynchronously(mediaPackage);
```

Comment:
This looks weird to me. AFAICT the media package was gotten from the asset manager snapshot so all element urls are archive urls, not the distribution urls???

Reply:
Yeah, that should definitely be checked.
Removal of old elements is a problem with our current merge workflow, too, though.